src/python/pants/backend/python/tasks/python_repl.py

```diff
@@ -46,8 +49,19 @@ def setup_repl_session(self, targets):
       entry_point = self.get_options().ipython_entry_point
     else:
       entry_point = 'code:interact'
+    # Merge interpreter constraints
+    repl_constraints = self.get_options().interpreter_compatibility
+    target_constraints = [constraint for target in targets for constraint in target.compatibility]
+    if repl_constraints and target_constraints:
+      interpreter_constraints = [",".join(pair) for pair in zip(repl_constraints, target_constraints)]
```

Comment:
Looks like merging will become a bit more intricate than first expected. We want a single string like you have here so that we AND the constraints, however we need to do things like de-duplicate interpreter class (`CPython`) and this will currently fail to parse: `CPython>=2.7,<3,CPython>3`. Also, this is a sizable chunk that would be a great opportunity for abstraction into a helper instance method, say `merge_interpreter_constraints(targets)` or similar.

Reply:
I will follow up with you about this offline.